### PR TITLE
Add VM Lifecycle Management Test

### PR DESCRIPTION
**Pull request by veedlaw**
Automated test case: xxxx-33025

Special Handling for UEFI:
- The test includes handling for UEFI guests, where the EFI vars NVRAM file is backed up before updating the VM's XML to avoid issues with UEFI VMs.

### CODE REVIEW

File: libvirt/tests/src/domain_life_cycle/on_xml_element_life_cycle.py [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+
+    IS_UEFI = None
+    temp_dir = None
+    source_file = None
+    backup_file = None
+
+    try:
+        # Test step 1: Check the guest can shutdown via virsh.shutdown()
+        LOG.info("Starting test step 1.")
+        if not vm.is_alive():
```
Comment (smitterl) [inline libvirt/tests/src/domain_life_cycle/on_xml_element_life_cycle.py:102]: The VM is started if 'start_vm = yes' in the test cfg, so you could avoid these two lines.

File: libvirt/tests/src/domain_life_cycle/on_xml_element_life_cycle.py [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+            vm.start()
+        session = vm.wait_for_login(timeout=LOGIN_WAIT_TIMEOUT)
+        IS_UEFI = utils_misc.is_linux_uefi_guest(lambda x: utils_misc.cmd_status_output(x, session=session)) or \
```
Comment (smitterl) [inline libvirt/tests/src/domain_life_cycle/on_xml_element_life_cycle.py:105]: You _could_ wrap this in a smaller function for readability.

File: libvirt/tests/src/domain_life_cycle/on_xml_element_life_cycle.py [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+        # Backup the efi vars file: (UEFI special edge case)
+        # To update a vm's xml we use vmxml.sync()
+        # sync() uses virsh.undefine and virsh.define in the backend
+        # virsh.undefine deletes the efi vars nvram file
+        # virsh.define does not "restore" it
+        # after starting the UEFI vm it sees the file is not populated and issues a reboot
+        # reboot=destroy, vm does not come up
+        # try to start the vm, goto step5
+        # to work around this we back up the file before sync and restore it after sync
+        source_file = vmxml.get_os().get_nvram()
```
Comment (smitterl) [inline libvirt/tests/src/domain_life_cycle/on_xml_element_life_cycle.py:125]: I would usually expect this level of detail in the commit message. In the code however, more high level (something like sync removes the nvram file which causes UEFI guests to reset/reboot).

File: libvirt/tests/src/domain_life_cycle/on_xml_element_life_cycle.py [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+        temp_dir = tempfile.TemporaryDirectory()
+        backup_file = pathlib.Path(temp_dir.name) / pathlib.Path(source_file).name
+
+        backup_success = backup_nvram_file(source_file, backup_file)
+        if not backup_success:
+            test.error("Backing up nvram file failed")
```
Comment (smitterl) [inline libvirt/tests/src/domain_life_cycle/on_xml_element_life_cycle.py:131]: Shorter: `if not backup_nvram_file(...)` - like you did with the `vm.shutdown` above.

File: libvirt/tests/src/domain_life_cycle/on_xml_element_life_cycle.py [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+    vmxml.set_on_reboot(params.get("on_reboot"))
+    LOG.debug("Set on_reboot to: %s" % vmxml.get_on_reboot())
+    vmxml.set_on_crash(params.get("on_crash"))
+    LOG.debug("Set on_crash to: %s" % vmxml.get_on_crash())
+    vmxml.sync()
+
+    if IS_UEFI:
+        restore_success = restore_nvram_file(backup_file, source_file)
+        if not restore_success:
+            test.error("Restoring nvram file failed")
```
Comment (smitterl) [inline libvirt/tests/src/domain_life_cycle/on_xml_element_life_cycle.py:144]: `if not restore_nvram...`

File: libvirt/tests/src/domain_life_cycle/on_xml_element_life_cycle.py [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+        test.error("VM failed initial shutdown")
+    try:
+        # Test step 2: "reboot" guest inside the guest OS
+        LOG.info("Starting test step 2.")
+        if not vm.is_alive():
+            LOG.debug("Step 2 starting vm")
+            vm.start()
+        session = vm.wait_for_login(timeout=LOGIN_WAIT_TIMEOUT)
+        try:
+            vm.reboot(method="shell", timeout=10)  # timeout = 10 because we don't want to login but wwait for the vm to go down
```
Comment (smitterl) [inline libvirt/tests/src/domain_life_cycle/on_xml_element_life_cycle.py:156]: typo s/wwait/wait/

File: libvirt/tests/src/domain_life_cycle/on_xml_element_life_cycle.py [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+        except (remote.LoginTimeoutError, virt_vm.VMDeadError):
+            # We are expecting this error and **must** ignore it otherwise it will halt the execution
+            LOG.debug(f'Step 2: Exception caught as expected')
+            pass
```
Comment (smitterl) [inline libvirt/tests/src/domain_life_cycle/on_xml_element_life_cycle.py:160]: Would `virsh.reboot` result in simpler code?

File: libvirt/tests/src/domain_life_cycle/on_xml_element_life_cycle.py [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+
+        if not utils_misc.wait_for(
+            lambda: not session.is_responsive(),
+                timeout=LOGIN_WAIT_TIMEOUT):
+            test.fail("Step 2: VM is still alive after console reboot with behaviour set to 'destroy'")
+        if vm.is_alive():
+            test.fail("VM is still alive after console reboot")
```
Comment (smitterl) [inline libvirt/tests/src/domain_life_cycle/on_xml_element_life_cycle.py:167]: Why do you test both `is_responsive` and `is_alive` here? Also, where do you confirm that the setting is 'destroy', are there other test case configs that would make this step fail? It looks like on one hand you are reading the behavior config, e.g. `vmxml.set_on_reboot(params.get("on_reboot"))` but then the rest of the code assumes that it's set to 'destroy' always. But I might misunderstand this.

File: libvirt/tests/src/domain_life_cycle/on_xml_element_life_cycle.py [hunk 9]
```diff
@@ -0,0 +1,199 @@
...
+                timeout=LOGIN_WAIT_TIMEOUT):
+            test.fail("Step 3: VM is still alive after reboot with behaviour set to 'destroy'")
+        LOG.info("Test step 3 passed.")
+
+        # Test step 4: "virsh shutdown <guest>"
+        LOG.info("Starting test step 4.")
+        vm.start()
+        session = vm.wait_for_login(timeout=LOGIN_WAIT_TIMEOUT)
+        if not virsh.shutdown(vm_name):
+            test.reboot("Shutdown failed")
```
Comment (smitterl) [inline libvirt/tests/src/domain_life_cycle/on_xml_element_life_cycle.py:187]: `test.reboot` looks like a typo.

**Review by smitterl — CHANGES_REQUESTED** [review]

**Comment by smitterl** [comment]
Please add the test results of a run in a comment.

File: libvirt/tests/src/domain_life_cycle/on_xml_element_life_cycle.py [hunk 10]
```diff
@@ -0,0 +1,199 @@
...
+    """
+    LOGIN_WAIT_TIMEOUT = float(params.get("login_timeout"))
+    vm_name = params.get("main_vm")
+    vm = env.get_vm(vm_name)
+    vmxml = vm_xml.VMXML.new_from_inactive_dumpxml(vm_name)
+    backup_xml = vmxml.copy()
+    session = None
+
+    IS_UEFI = None
+    temp_dir = None
```
[the remark this answers is not in the document]
Comment (veedlaw) [reply libvirt/tests/src/domain_life_cycle/on_xml_element_life_cycle.py:95]: Whilst it is not a constant in the strict sense, it was named in uppercase letters to help signal that it's a variable that behaves as a constant throughout the code.  It is declared as None at the start, because the correct value depends on the vm session being started inside the try-block.  I considered it a helpful indicator, but I'll change it if it is confusing.

File: libvirt/tests/src/domain_life_cycle/on_xml_element_life_cycle.py (hunk 7 above)
Comment (veedlaw) [reply libvirt/tests/src/domain_life_cycle/on_xml_element_life_cycle.py:160]: The difference between step 2 and step 3 of the test is:
Step 2: "reboot" inside guest os
Step 3: use "virsh reboot" <guest>

File: libvirt/tests/src/domain_life_cycle/on_xml_element_life_cycle.py (hunk 8 above)
Comment (veedlaw) [reply libvirt/tests/src/domain_life_cycle/on_xml_element_life_cycle.py:167]: You are right, checking vm.is_alive() is redundant here.
The xml to be set to destroy is defined in test case xxxx-33025.

**Comment by veedlaw** [comment]
Test run results:
```
# avocado run --vt-type libvirt --test-runner=runner my-example-test.destroy
No python imaging library installed. Screendump and Windows guest BSOD detection are disabled. In order to enable it, please install python-imaging or the equivalent for your distro.
No python imaging library installed. PPM image conversion to JPEG disabled. In order to enable it, please install python-imaging or the equivalent for your distro.
JOB ID     : c35459111080439ee4aa7a7ca86d3415371ec731
JOB LOG    : /var/lib/avocado/job-results/job-2024-02-27T07.20-c354591/job.log
 (1/1) type_specific.io-github-autotest-libvirt.my-example-test.destroy: PASS (298.15 s)
RESULTS    : PASS 1 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /var/lib/avocado/job-results/job-2024-02-27T07.20-c354591/results.html
JOB TIME   : 324.19 s
```

```
# avocado run --vt-type libvirt --test-runner=runner my-example-test.coredump_destroy
No python imaging library installed. Screendump and Windows guest BSOD detection are disabled. In order to enable it, please install python-imaging or the equivalent for your distro.
No python imaging library installed. PPM image conversion to JPEG disabled. In order to enable it, please install python-imaging or the equivalent for your distro.
JOB ID     : 2a8c317d7dfc091f722c72679de66508f4f4dd30
JOB LOG    : /var/lib/avocado/job-results/job-2024-02-27T07.39-2a8c317/job.log
 (1/1) type_specific.io-github-autotest-libvirt.my-example-test.coredump_destroy: PASS (287.89 s)
RESULTS    : PASS 1 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /var/lib/avocado/job-results/job-2024-02-27T07.39-2a8c317/results.html
JOB TIME   : 313.87 s

```

File: libvirt/tests/src/domain_life_cycle/on_xml_element_life_cycle.py (hunk 7 above)
Comment (smitterl) [reply libvirt/tests/src/domain_life_cycle/on_xml_element_life_cycle.py:150]: Thanks for clarifying.

File: libvirt/tests/src/domain_life_cycle/on_xml_element_life_cycle.py (hunk 2 above)
Comment (smitterl) [reply libvirt/tests/src/domain_life_cycle/on_xml_element_life_cycle.py:103]: I find the code really hard to read. As everything is written in a single function (`run`) all abstraction levels are in the same code. By clean code principles each function should strive to maintain the same level of abstraction ref. https://www.code4it.dev/cleancodetips/coherent-levels-of-abstraction/

As a result, if you want to improve the code, I recommend:

1. move your steps into functions
2. give the functions the corresponding name of the step, then there's no need to add a comment anymore which might more easily become out of sync of the actual code, another clean code principle, ref. https://dev.to/gervaisamoah/a-guide-to-clean-code-the-power-of-good-names-3f6i
3. IMO there's not need to log start and end of the steps, but if you prefer it this way, you can write that more efficient if you use those functions and the decorator pattern, you can even use python's decorator implementation
```python
def run(...):
...
    run_step("1", lambda: initial_shutdown(...))
    run_step("2", lambda: reboot_inside_guest(...))
...
def run_step(number, test_func):
    LOG.info("Start step %s", number)
    test_func()
    LOG.info("Stop step %s", number)
```

File: libvirt/tests/src/domain_life_cycle/on_xml_element_life_cycle.py [hunk 11]
```diff
@@ -0,0 +1,187 @@
...
+        # Updating a VM's XML using vmxml.sync() deletes the nvram file, which on a UEFI VM
+        # causes an unwanted reset/reboot.
+        source_file = vmxml.get_os().get_nvram()
+        temp_dir = tempfile.TemporaryDirectory()
+        backup_file = pathlib.Path(temp_dir.name) / pathlib.Path(source_file).name
+
+        if not backup_nvram_file(source_file, backup_file):
+            test.error("Backing up nvram file failed")
+
+    vmxml.set_on_poweroff(params.get("on_poweroff"))
```
Comment (smitterl) [inline libvirt/tests/src/domain_life_cycle/on_xml_element_life_cycle.py:124]: We usually, shorter, just print the VM XML after the update
```
LOG.debug("New VMXML: ", vm_xml.VMXML.new_from_inactive_dumpxml(...))
```
The updated XML is generally useful.

File: libvirt/tests/src/domain_life_cycle/on_xml_element_life_cycle.py [hunk 12]
```diff
@@ -0,0 +1,187 @@
...
+        session = vm.wait_for_login(timeout=LOGIN_WAIT_TIMEOUT)
+        IS_UEFI = utils_misc.is_linux_uefi_guest(lambda x: utils_misc.cmd_status_output(x, session=session)) or \
+            utils_misc.is_windows_uefi_guest(lambda x: utils_misc.cmd_status_output(x, session=session))
+        LOG.debug("VM is UEFI: %s", IS_UEFI)
+        if not vm.shutdown():
+            test.error("VM failed initial shutdown")
+        LOG.info("Test step 1 passed.")
+    finally:
+        if session:
+            session.close()
```
Comment (smitterl) [inline libvirt/tests/src/domain_life_cycle/on_xml_element_life_cycle.py:111]: Here you close the session whereas further down you don't, is there a reason for it?

**Review by smitterl — CHANGES_REQUESTED** [review]
Thanks for this. I find the code quality should be improved. I found it really hard to review. Please check.

File: libvirt/tests/src/domain_life_cycle/on_xml_element_life_cycle.py (hunk 11 above)
Comment (yalzhang) [reply libvirt/tests/src/domain_life_cycle/on_xml_element_life_cycle.py:124]: This may be unnecessary once https://github.com/avocado-framework/avocado-vt/pull/3955/files got merged.
